src/dmd/dinterpret.d

```diff
@@ -4878,6 +4878,10 @@ public:
 
                 return;
             }
+            else if (fd.ident == Id._d_arrayappendT)
+                assert(0, "CTFE cannot interpret _d_arrayappendT!");
+            else if (fd.ident == Id._d_arrayappendcTX)
+                assert(0, "CTFE cannot interpret _d_arrayappendcTX!");
```

Comment:
Having another look over, if I understand this function right, you should instead implement these two functions before CTFE _can_ interpret arrayappend operations.  Then there's no need for the `(__ctfe) ? a : b` set-up you've done in semantic, just generate the lowering to template directly.

Reply:
If the lowering reaches CTFE, this error [1] is triggered because of the mixin in the hook's body [2]. That's why those `assert`s are here and that's why I use `__ctfe ? ...`.

[1] https://github.com/dlang/dmd/blob/b14861d42ce3da9cf789988621c46dba1102ee0c/src/dmd/dinterpret.d#L4967-L4972
[2] https://github.com/dlang/druntime/blob/cc101972b1f6d3047ca7b54cbf4aee11ac499b75/src/core/internal/array/appending.d#L42